palette/src/matrix.rs

```diff
@@ -42,58 +56,66 @@ pub fn multiply_rgb_to_xyz<S: RgbSpace, T: FloatComponent>(
     c: &Mat3<T>,
     f: &Rgb<Linear<S>, T>,
 ) -> Xyz<S::WhitePoint, T> {
+    let [c0, c1, c2, c3, c4, c5, c6, c7, c8] = *c;
+
     Xyz {
-        x: (c[0] * f.red) + (c[1] * f.green) + (c[2] * f.blue),
-        y: (c[3] * f.red) + (c[4] * f.green) + (c[5] * f.blue),
-        z: (c[6] * f.red) + (c[7] * f.green) + (c[8] * f.blue),
+        x: (c0 * f.red) + (c1 * f.green) + (c2 * f.blue),
+        y: (c3 * f.red) + (c4 * f.green) + (c5 * f.blue),
+        z: (c6 * f.red) + (c7 * f.green) + (c8 * f.blue),
         white_point: PhantomData,
     }
 }
 
 /// Multiply two 3x3 matrices.
 pub fn multiply_3x3<T: Float>(c: &Mat3<T>, f: &Mat3<T>) -> Mat3<T> {
-    let mut out = [T::zero(); 9];
-    out[0] = c[0] * f[0] + c[1] * f[3] + c[2] * f[6];
-    out[1] = c[0] * f[1] + c[1] * f[4] + c[2] * f[7];
-    out[2] = c[0] * f[2] + c[1] * f[5] + c[2] * f[8];
+    let [c0, c1, c2, c3, c4, c5, c6, c7, c8] = *c;
+    let [f0, f1, f2, f3, f4, f5, f6, f7, f8] = *f;
+
+    let o0 = c0 * f0 + c1 * f3 + c2 * f6;
+    let o1 = c0 * f1 + c1 * f4 + c2 * f7;
+    let o2 = c0 * f2 + c1 * f5 + c2 * f8;
 
-    out[3] = c[3] * f[0] + c[4] * f[3] + c[5] * f[6];
-    out[4] = c[3] * f[1] + c[4] * f[4] + c[5] * f[7];
-    out[5] = c[3] * f[2] + c[4] * f[5] + c[5] * f[8];
+    let o3 = c3 * f0 + c4 * f3 + c5 * f6;
+    let o4 = c3 * f1 + c4 * f4 + c5 * f7;
+    let o5 = c3 * f2 + c4 * f5 + c5 * f8;
 
-    out[6] = c[6] * f[0] + c[7] * f[3] + c[8] * f[6];
-    out[7] = c[6] * f[1] + c[7] * f[4] + c[8] * f[7];
-    out[8] = c[6] * f[2] + c[7] * f[5] + c[8] * f[8];
+    let o6 = c6 * f0 + c7 * f3 + c8 * f6;
+    let o7 = c6 * f1 + c7 * f4 + c8 * f7;
+    let o8 = c6 * f2 + c7 * f5 + c8 * f8;
 
-    out
+    [o0, o1, o2, o3, o4, o5, o6, o7, o8]
 }
 
 /// Invert a 3x3 matrix and panic if matrix is not invertible.
 pub fn matrix_inverse<T: Float>(a: &Mat3<T>) -> Mat3<T> {
+    assert!(a.len() > 8);
```

Comment:
It's probably good to add a comment here to explain why it's different. The same in other places where there's a risk of someone coming by and "fixing" them. 🙂